add column types for IPv4 and IPv6 addresses, compatible with both text and numeric filters, thanks to natural sort

kirisakow opened the pull request:
This is an attempt to respond to issue #402 (and #981) by implementing `TextIPv4Address` and `TextIPv6Address` column types that are compatible with both text and numeric `FilterIds` (all nine of them), thanks to natural sort.

This is my first PR on this project, so please feel free to kindly point out anything I may have missed or done wrong.

kirisakow committed:
sort column types

kirisakow committed:
add ColumnType TextIPv4Address

kirisakow committed:
add ColumnType TextIPv6Address

kirisakow committed:
fix: column type-subtype value may not contain more than one hyphen

kirisakow committed:
add radio buttons for Create Column modal dialog

kirisakow committed:
l10n: add placeholders for labels "IPv4 address" and "IPv6 address"

kirisakow committed:
add column type partials

kirisakow committed:
rename as per the declared column types

enjeck commented:
@kirisakow Thanks for working on this. Coincidentally, I've also been working on a new column type (for users/groups) at: https://github.com/nextcloud/tables/pull/1090. Maybe the files and feedback I'm getting over there helps :smile:

enjeck commented:
@kirisakow While working on https://github.com/nextcloud/tables/pull/1090, I wrote down some notes of files I potentially had to change/add. I think not everything is applicable here, but might be helpful:

### Frontend
- In `src/shared/components/ncTable/mixins/columnsTypes`, we create a new column type that specifies:
    - properties the column has
   - converting value to string representation (and using for search and filter operations)
   - sort function
- In `src/shared/components/ncTable/mixins/magicFields.js`, we specify if the new column fits any particular magicfield
- In `src/shared/components/ncTable/mixins/filters.js`, we specify if the new column should be shown for any particular filter
- In `src/shared/components/ncTable/mixins/columnParser.js`, we specify that this new column is valid
- In `src/shared/components/ncTable/mixins/columnHandler.js`, add the column and subtypes if any
- In `src/modules/modals/EditColumn.vue` and `src/modules/modals/CreateColumn.vue`, specify how to handle column creation and editing
- in `src/modules/main/partials/ColumnFormComponent.vue`, add the column
- In `src/modules/main/partials/ColumnTypeSelection.vue`, add icon and text that will appear when user initially chooses a column type
- In `src/shared/constants.js` , specify new column type
- In `src/shared/components/ncTable/mixins/columnClass.js`, maybe add Abstract column
- In `src/shared/components/ncTable/mixins/columnHandler.js`, add datatype
- In `src/shared/components/ncTable/mixins/columnParser.js`, add type
- In `src/shared/components/ncTable/mixins/metaColumns.js`, maybe add meta type
- In `src/shared/components/ncTable`, add new TableCell component file for rendering the type in the table.
- In `src/shared/components/ncTable/partials/columnTypePartials/forms`, add a new Form component that will be inserted into the create/edit form when we try to create a new column for this new type. Will have to create this for subtypes too
- In `src/shared/components/ncTable/partials/rowTypePartials`, add a new Form component that will be inserted into the create/edit form when we try to create a new row for this new type.  Will have to create this for subtypes too, if there are any

### Backend
- In `lib/Service/ColumnService.php`, specify properties
- In `lib/Service/ColumnTypes`, add a new column type
- In `lib/Db/Column.php, add new column type
- Add column properties to `lib/Service/TableTemplateService.php`
- Add to `lib/Controller/ColumnController.php` the properties for new type
- Add type to `lib/Controller/ApiColumnsController.php`
- Add column details to `lib/ResponseDefinitions.php`
- Add column type to `lib/Capabilities.php`
- Add route to create column to `appinfo/routes.php`
- Add properties to `lib/Controller/Api1Controller.php`
- Add new column type file to `lib/Db/ColumnTypes`
- Add to new column data `lib/Db/LegacyRowMapper.php`
- Add to `lib/Db/Row2Mapper.php`
- Add new RowCell and RowCellMapper php file for the new column type to `lib/Db`
- Add column type to `lib/Helper/ColumnsHelper.php`
- Update migration files to include new type. Look at all files at `lib/Migration` most especially `lib/Migration/Version000000Date20210921000000.php` and `lib/Migration/Version000700Date20230916000000.php`
- Maybe add to `lib/Service/ImportService.php`

### Others
- Ensure types are added to `src/types/openapi/openapi.ts`
- Add bdd test to `tests/integration/features/APIv2.feature`
- Ensure types are added to `openapi.json`
- Write e2e tests

kirisakow committed:
use built-in JS func `localeCompare` to compare two strings.

kirisakow committed:
update magic fields

kirisakow committed:
rename class name: append `Column` suffix

kirisakow committed:
update `columnParser.js`

kirisakow committed:
update `EditColumn.vue`

kirisakow committed:
update `ColumnFormComponent.vue`

kirisakow committed:
fix: CSS-compatible comments syntax

kirisakow committed:
add row type partials

kirisakow committed:
sort imports

kirisakow committed:
update `CreateColumn.vue`

kirisakow committed:
sort imports

kirisakow commented:
@enjeck Thank you for the input, that was of much help!

I assume development is now finished for the
- [x] front-end part;
- [x] back-end part;

Still, a few TODOs remain — like things I wasn't sure about:
- [ ] specify [valid IPv4 and IPv6 regex patterns][1], as I wasn't sure where and how to declare a value of `textAllowedPattern`;
- [ ] every thing tests-related — unit tests, integration tests;
- [ ] every thing OpenAPI-related*;
- [ ] every thing migrations-related*;

\* probably nothing to do about OpenAPI or migrations, as this feature merely adds two subtypes of the existing `text` type.

[1]: https://stackoverflow.com/questions/23483855/javascript-regex-to-validate-ipv4-and-ipv6-address-no-hostnames/69685444#69685444

kirisakow committed:
fix: undo a misfocused search pattern that was baked in to the committed code

kirisakow committed:
fix: `Module not found: Error`

kirisakow committed:
fix: `Error: Form ... does no exist` by changing subtype value declared in `columnHandler.js` and its occurrencies in `CreateColumn.vue`

kirisakow committed:
fix: `Error: Form ... does no exist`:
- change subtype value declared in `columnHandler.js` and its occurrencies in `CreateColumn.vue`;
- rename the corresponding .vue files;
- refactor the corresponding imports.

kirisakow committed:
fix: default component name

kirisakow commented:
@enjeck Hi guys, it's been a long while since I quit this PR. Sorry about that. If I remember well, I quit after I had to accept I couldn't manage to go all the way down all the PHP layers to make it work.

Thanks for reactivating this PR. This would still be a great new feature for the Tables app if you guys could finish the job and make it work.

enjeck commented:
> @enjeck Hi guys, it's been a long while since I quit this PR. Sorry about that. If I remember well, I quit after I had to accept I couldn't manage to go all the way down all the PHP layers to make it work.
> 
> Thanks for reactivating this PR. This would still be a great new feature for the Tables app if you guys could finish the job and make it work.

Thank you for contributing @kirisakow . I'll see what I can do. If it turns out that we can use `textAllowedPattern` to regex match patterns, then we can use that for supporting IPV4 and IPV6, and avoid making too many changes

enjeck commented:
easier done using textAllowedPattern, as explained at https://github.com/nextcloud/tables/issues/981#issuecomment-4830435795
